_cite/cite.py

```diff
@@ -63,37 +64,43 @@
 
             # run plugin on data entry to expand into multiple sources
             try:
-                entry = import_module(f"plugins.{plugin.stem}").main(entry)
+                expanded = import_module(f"plugins.{plugin.stem}").main(entry)
                 # check that plugin returned correct format
-                if not list_of_dicts(entry):
+                if not list_of_dicts(expanded):
                     raise Exception("Plugin didn't return list of dicts")
             except Exception as e:
+                print(traceback.format_exc())
                 log(e, 3, "ERROR")
                 error = True
                 continue
 
             # loop through sources
-            for source in entry:
+            for index, source in enumerate(expanded):
                 if plugin.stem != "sources":
                     log(label(source), 3)
 
                 # include meta info about source
                 source["plugin"] = plugin.name
                 source["file"] = file.name
-                # add source to master list
+
+                # add source to compiled list
                 sources.append(source)
 
             if plugin.stem != "sources":
-                log(f"{len(entry)} source(s)", 3)
+                log(f"{len(expanded)} source(s)", 3)
+
 
+log("Merging sources by id")
 
 # merge sources with matching (non-blank) ids
 for a in range(0, len(sources)):
-    _id = sources[a].get("id", "")
-    if not _id:
+    a_id = get_safe(sources, str(a) + ".id", "")
+    if not a_id:
         continue
     for b in range(a + 1, len(sources)):
-        if sources[b].get("id", "") == _id:
+        b_id = get_safe(sources, str(b) + ".id", "")
+        if b_id == a_id:
+            log(f"Found duplicate {b_id}", 2)
             sources[a].update(sources[b])
             sources[b] = {}
 sources = [entry for entry in sources if entry]
```

Comment:
Going off of @d33bs' comment above about `zip()` being better than iterating over indices, it looks like what you're doing here is kind of like bubble sort, except that you're collapsing entries with the same ID into a single entry in the result. You could use Python's `Counter` and `defaultdict`, both part of Python's standard library, for the same purpose of detecting and merging duplicates, e.g.:

```python
from collections import Counter, defaultdict

# produces a list of IDs for which duplicates were detected
duplicates = [
  k for k, v in Counter(x.get("id") for x in sources).items() if v and v > 1
]

print(f"Duplicated IDs in sources: {", ". join(duplicates)}")

# merges elements with the same ID into a single element (if the element has an ID).
# a defaultdict is like a dict, but instead of throwing KeyError when you access a non-
#  existent key it uses the factory method with which it's initialized (in this case, dict)
#  to produce a new value at that key.
# a.update(b) will effectively set 'a' to 'b' if 'a' is an empty dict (which it will be on first
#  access with defaultdict), so this is safe.
source_dict = defaultdict(dict)
for source in [x for x in sources if x.get("id") is not None]:
  source_dict[source.get("id")].update(source)
sources = list(source_dict.values())
```

Reply:
As is, this code seems to filter out entries without an `id` from `sources`, so it doesn't meet the requirements.

Reply:
Well, if you're happy with the code as-is and it works, that's fine, of course. Sorry, didn't realize on reading your code that that's what it did, and you're right that keeping the sources list in the same order while merging duplicates would be a little verbose with my approach.

Reply:
Ah yes, order is something I forgot about too. If someone writes them in a specific order in sources, they probably want to keep that order when using the list component to display them. I don't see any clearer way to do it other than nested for loops. It might not be pythonic, but imo two nested for loop is a standard and well-recognized way to do a triangular matrix type of thing.